tests/InferTypesFuzz.elm

```diff
@@ -0,0 +1,329 @@
+            starters ++ others
+
+        charFrom string =
+            string
+                |> String.toList
+                |> List.map Random.constant
+                |> Random.choices (Random.constant 'a')
+
+        firstChar =
+            charFrom starters
+
+        rest =
+            Random.list 5 <| charFrom all
+    in
+    Random.map2 (::) firstChar rest
+        |> Random.map String.fromList
+        |> Random.map VarName
+
+
+dumpType : Type -> String
+dumpType type_ =
+    type_
+        |> Type.toString Type.emptyState
+        |> Tuple.first
+
+
+{-| An expression shrinker that preserves the inferred type.
+-}
+exprShrinker : Shrinker Canonical.LocatedExpr
+exprShrinker expr =
+    case Located.unwrap expr of
+        Canonical.Literal lit ->
+            lit
+                |> shrinkLiteral
+                |> Shrink.map Canonical.Literal
+                |> Shrink.map located
+
+        Canonical.Plus left right ->
+            expr
+                |> shrinkPlus left right
+
+        Canonical.List elements ->
+            elements
+                |> shrinkNonEmptyList exprShrinker
+                |> Shrink.map Canonical.List
+                |> Shrink.map located
+
+        Canonical.If { then_, else_ } ->
+            -- TODO: Shrink the test once we get more interesting Bool expressions.
+            [ then_, else_ ]
+                |> List.map shrinkTo
+                |> concatShrink
+                |> shrink expr
+
+        Canonical.Lambda { argument, body } ->
+            expr
+                |> shrinkLambda argument body
+
+        _ ->
+            expr |> Shrink.noShrink
+
+
+shrinkLiteral : Shrinker Literal
+shrinkLiteral lit =
+    case lit of
+        Int i ->
+            i |> Shrink.int |> Shrink.map Int
+
+        Float f ->
+            f |> Shrink.float |> Shrink.map Float
+
+        Char c ->
+            c |> Shrink.char |> Shrink.map Char
+
+        Bool b ->
+            b |> Shrink.bool |> Shrink.map Bool
+
+        String s ->
+            s |> Shrink.string |> Shrink.map String
+
+
+shrinkPlus : Canonical.LocatedExpr -> Canonical.LocatedExpr -> Shrinker Canonical.LocatedExpr
+shrinkPlus left right _ =
+    lazyMap2 Canonical.Plus
+        (exprShrinker left)
+        (exprShrinker right)
+        |> Shrink.map located
+
+
+shrinkLambda : VarName -> Canonical.LocatedExpr -> Shrinker Canonical.LocatedExpr
+shrinkLambda argument body _ =
+    body
+        |> exprShrinker
+        |> Shrink.map (Canonical.lambda argument)
+        |> Shrink.map located
+
+
+shrinkNonEmptyList : Shrinker a -> Shrinker (List a)
+shrinkNonEmptyList shrinkElement list =
+    case list of
+        first :: rest ->
+            lazyMap2 (::)
+                (shrinkElement first)
+                (Shrink.list shrinkElement rest)
+
+        [] ->
+            Shrink.noShrink list
+
+
+shrinkTo : a -> Shrinker a
+shrinkTo shrunk _ =
+    True
+        |> Shrink.bool
+        |> Shrink.map (always shrunk)
```

Comment:
I don't get what this function does. Is it something `Random.constant`-like?
If yes, does that mean that `If` shrinks like this?
```elm
If { test = _, then_ = a, else_ = b }
--> [ a, b ]
```
Does it shrink the `a` and `b` too? It seems to me like it doesn't but I might have missed something.

Reply:
Now that you've pointed it out I think the other cases might be wrong. AFAIR shrinkers should only do one step of shrinking at a time. elm-test is responsible for shrinking the output again as long it's possible. :man_facepalming: